### ISSUE TITLE
Incorrectly asserting an object is a type instead of using isinstance

### ISSUE THREAD

**Issue by MichaelMallaburn** [astral-sh/ty#3878]
### Summary

Using Python 3.14.6 and ty 0.0.55, and given the following python file:
```python
#!/usr/bin/env python3

def takes_int(a: int) -> None:
    pass


def use_maybe_int1(a: int | None) -> None:
    assert a is int  # Wrong assertion.
    takes_int(a)


def use_maybe_int2(a: int | None) -> None:
    assert type(a) is int
    takes_int(a)


def use_maybe_int3(a: int | None) -> None:
    assert isinstance(a, int)
    takes_int(a)
```
, and a ty.toml of
```toml
# https://docs.astral.sh/ty/reference/configuration/
[environment]
root = ["./src"]

[src]
exclude = [".pixi"]
```
, the command `ty check` passes with the message
> All checks passed!

The issue is that `a: int | None` is incompatible with `assert a is int`.
Because execution ends at the assertion, no further check is done by ty, and it does not report the following call to `takes_int`.

Maybe the fix is to produce a warning for an assertion against a type that will always produce a failure.
Hopefully that can be done without triggering a lot of noise.

Thank you

### Version

0.0.55

**Comment by sharkdp** [comment]
Thank you for reporting this.

> The issue is that `a: int | None` is incompatible with `assert a is int`. Because execution ends at the assertion, no further check is done by ty, and it does not report the following call to `takes_int`.

I think problems of this type will generally be solved by implementing a rule like #1948 which would notify you about blocks of code that are unreachable.

> Maybe the fix is to produce a warning for an assertion against a type that will always produce a failure.

Yes. A more general idea would be to emit a diagnostic on assertion conditions that ty infers as statically `True` or `False`. This is tracked in https://github.com/astral-sh/ty/issues/3401.

Given this, I'm not sure if there is a lot of value in implementing a special diagnostic that would detect this specific `a is b` pattern with incompatible types of `a` and `b`. I'm closing this for now, but if someone thinks that should also be a rule, let me know and we can reopen this.
